File: lexer.cpp
#include <iostream>
#include <vector>
#include <regex>
#include <string>

enum class TokenType{
    Identifier,
    Number,
    Operator,
    Unknown
};

struct Token
{
    std::string value;
    TokenType type;
};

struct  TokenRule
{
    TokenType type;
    std::regex pattern;
};


class Lexer{
    private:
        std::string source;
        size_t pos;
    public:
        Lexer(const std::string &src){
            source = src;
            pos = 0;
        }

        std::vector<Token> tokenize(){
            std::vector<Token> tokens;

            std::vector<TokenRule> rules ={
                {TokenType::Identifier, std::regex("^[A-Za-z_][A-Za-z0-9_]*")},
                {TokenType::Number, std::regex("^[0-9]+")},
                {TokenType::Operator, std::regex("^(=|\\+|\\-|\\*|/)")}
            };
            
            std::string input = source;

            while (!input.empty())
            {
                bool matched = false;

                for(auto& rule : rules){
                    std::smatch match;
                    if(std::regex_search(input, match, rule.pattern)){
                        tokens.push_back({match.str(), rule.type});
                        input = input.substr(match.length());
                        matched = true;
                        break;
                    }
                }

                if(!matched){
                    tokens.push_back({std::string(1, input[0]), TokenType::Unknown});
                    input = input.substr(1);
                }
            }
            
            return tokens;
        };
};

File: JavaCompiler/frontend/Lexer.java

package frontend;

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

import java.util.regex.Matcher;


public class Lexer {
    String src;
    long pos;

    public Lexer(String src){
        this.src = src;
        pos = 0;
    }

    public ArrayList<Token> Tokenizer(){
        ArrayList<Token> tokens = new ArrayList<>();
        
        List<TokenRule> rules = Arrays.asList(
            new TokenRule(TokenTypes.Identifier, "^[A-za-z_][A-za-z0-9_]*"),
            new TokenRule(TokenTypes.Number, "^[0-9]+"),
            new TokenRule(TokenTypes.BinaryOperator, "^(=|\\+|\\-|\\*|/)")
        );

        String input = src;

        while(!input.isEmpty()){
            boolean matched = false;
            for(TokenRule rule : rules){
                Matcher matcher = rule.pattern.matcher(input);
                if(matcher.lookingAt()){
                    String value = matcher.group();
                    tokens.add(new Token(value, rule.type));
                    input = input.substring(matcher.end());
                    matched = true;
                    break;
                }
            }

            if(!matched){
                // IGNORE SPACES FOR NOW OR SOMETHING UNKNOWS IGNORE
                input = input.substring(1);
            }
        }
        return tokens;
    }
}

